pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndex.java

```diff
@@ -186,43 +186,54 @@ private RoaringBitmap getMatchingFlattenedDocIds(FilterContext filter) {
-    Preconditions.checkState(lhs.getType() == ExpressionContext.Type.IDENTIFIER,
+    Preconditions.checkArgument(lhs.getType() == ExpressionContext.Type.IDENTIFIER,
         "Left-hand side of the predicate must be an identifier, got: %s (%s). Put double quotes around the identifier if needed.",
         lhs, lhs.getType());
     String key = lhs.getIdentifier();
 
+    // Support 2 formats:
+    // - JSONPath format (e.g. "$.a[1].b"='abc', "$[0]"=1, "$"='abc')
+    // - Legacy format (e.g. "a[1].b"='abc')
+    if (key.charAt(0) == '$') {
+      key = key.substring(1);
+    } else {
+      key = JsonUtils.KEY_SEPARATOR + key;
+    }
+
     // Process the array index within the key if exists
-    // E.g. "foo[0].bar[1].foobar"='abc' -> foo.$index=0 && foo.bar.$index=1 && foo.bar.foobar='abc'
+    // E.g. "[*]"=1 -> "."='1'
+    // E.g. "[0]"=1 -> ".$index"='0' && "."='1'
+    // E.g. "[0][1]"=1 -> ".$index"='0' && "..$index"='1' && ".."='1'
+    // E.g. ".foo[*].bar[*].foobar"='abc' -> ".foo..bar..foobar"='abc'
+    // E.g. ".foo[0].bar[1].foobar"='abc' -> ".foo.$index"='0' && ".foo..bar.$index"='1' && ".foo..bar..foobar"='abc'
+    // E.g. ".foo[0][1].bar"='abc' -> ".foo.$index"='0' && ".foo..$index"='1' && ".foo...bar"='abc'
     RoaringBitmap matchingDocIds = null;
     int leftBracketIndex;
-    while ((leftBracketIndex = key.indexOf('[')) > 0) {
-      int rightBracketIndex = key.indexOf(']');
-      Preconditions.checkState(rightBracketIndex > leftBracketIndex, "Missing right bracket in key: %s", key);
+    while ((leftBracketIndex = key.indexOf('[')) >= 0) {
+      int rightBracketIndex = key.indexOf(']', leftBracketIndex + 2);
+      Preconditions.checkArgument(rightBracketIndex > 0, "Missing right bracket in key: %s", key);
 
       String leftPart = key.substring(0, leftBracketIndex);
-      int arrayIndex;
-      try {
-        arrayIndex = Integer.parseInt(key.substring(leftBracketIndex + 1, rightBracketIndex));
-      } catch (Exception e) {
-        throw new IllegalStateException("Invalid key: " + key);
-      }
+      String arrayIndex = key.substring(leftBracketIndex + 1, rightBracketIndex);
       String rightPart = key.substring(rightBracketIndex + 1);
 
-      // foo[1].bar -> foo.$index=1
-      String searchKey =
-          leftPart + JsonUtils.KEY_SEPARATOR + JsonUtils.ARRAY_INDEX_KEY + BaseJsonIndexCreator.KEY_VALUE_SEPARATOR
-              + arrayIndex;
-      RoaringBitmap docIds = _postingListMap.get(searchKey);
-      if (docIds != null) {
-        if (matchingDocIds == null) {
-          matchingDocIds = docIds.clone();
+      if (!arrayIndex.equals(JsonUtils.WILDCARD)) {
+        // "[0]"=1 -> ".$index"='0' && "."='1'
+        // ".foo[1].bar"='abc' -> ".foo.$index"=1 && ".foo..bar"='abc'
+        String searchKey = leftPart + JsonUtils.ARRAY_INDEX_KEY + BaseJsonIndexCreator.KEY_VALUE_SEPARATOR + arrayIndex;
+        RoaringBitmap docIds = _postingListMap.get(searchKey);
```

Comment:
Can I suggest using a different type for the key of the `_postingListMap` here and change the key type to some kind of tuple? It won't make any difference in this particular case because the strings making up `searchKey` are all new strings, but down on L233/244 the input strings may not have been modified. The problem with concatenating strings to create a key for a `HashMap` lookup is that they may be already have their hashcode computed and cached, making the hashcode of the concatentation very cheap to compute (just some arithmetic with precomputed hashcodes), but creating a new `String` means the hash code will need to be computed again. 

The allocation of the tuple will also be a lot smaller than constructing a new `String`, but with enough inlining the tuple will typically be scalarised, eliminating its allocation.

Reply:
Are you suggesting storing tuples like `[leftPart, JsonUtils.ARRAY_INDEX_KEY, BaseJsonIndexCreator.KEY_VALUE_SEPARATOR, arrayIndex]` and cache the hashcode of each part?
I can see that it might be able to reuse some strings and reduce some hashcode calculation, but it also has the following potential fallbacks:
- The tuple will be mutable, so maintaining the hashcode cache won't be easy. The extra logic of maintaining the cache might bring other overhead
- The tuple comparison logic won't be very straight forward. It might have to end up concatenate the strings in the tuple, or split the query string in order to compare them, which eliminate the benefit of the tuple
- The code won't be as readable as the string concatenation

This optimization is out of the scope of this PR. We can have a separate PR implementing this optimization if this part of the code become the hotspot

Reply:
I wouldn't suggest making the tuple type mutable, no. The `String` class caches its own hash code so there is no cache to maintain, so no added complexity. I typically observe an integer multiple reduction in latency of hash map lookups when making this transformation, because it makes the lookup genuinely O(1) rather than O(k) where k is the length of the concatenated string. As mentioned, this would also reduce allocation per lookup.

I agree this is out of the scope of the PR since it is an existing pattern, and that there is no profiling data to refer to in order to motivate the change, moreover, the cost of the lookup will be dwarfed by the subsequent bitmap intersection so you might not see it with a sampling profiler, but this looked like an efficiency smell when I was browsing this PR.

Reply:
(i.e. don't let this comment block merging the PR)

Reply:
I see. Yes, this is a good optimization, especially for `IN` and `NOT IN` predicate where the key is the same and only the value changes.
Since this is out of the scope of this PR, I added a TODO in the code and we can address it in a separate PR.